codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServiceConfigGenerator.kt

```diff
@@ -125,20 +243,199 @@ class ServiceConfigGenerator(
                         layers: self.layers,
                         http_plugins: self.http_plugins,
                         model_plugins: #{PluginStack}::new(model_plugin, self.model_plugins),
+                        #{BuilderRequiredMethodFlagsMove3:W}
+                    }
+                }
+                
+                #{BuilderBuildMethod:W}
+            }
+            """,
+            *codegenScope,
+            "BuilderRequiredMethodFlagsInit" to builderRequiredMethodFlagsInit(),
+            "BuilderRequiredMethodFlagDefinitions" to builderRequiredMethodFlagsDefinitions(),
+            "BuilderRequiredMethodError" to builderRequiredMethodError(),
+            "InjectedMethods" to injectedMethods(),
+            "BuilderRequiredMethodFlagsMove1" to builderRequiredMethodFlagsMove(),
+            "BuilderRequiredMethodFlagsMove2" to builderRequiredMethodFlagsMove(),
+            "BuilderRequiredMethodFlagsMove3" to builderRequiredMethodFlagsMove(),
+            "BuilderBuildMethod" to builderBuildMethod(),
+        )
+    }
+
+    private fun isBuilderFallible() = configMethods.isBuilderFallible()
+
+    private fun builderBuildRequiredMethodChecks() = configMethods.filter { it.isRequired }.map {
+        writable {
+            rust(
+                """
+                if !self.${it.requiredBuilderFlagName()} {
+                    return Err(${serviceName}ConfigError::${it.requiredErrorVariant()});
+                }
+                """,
+            )
+        }
+    }.join("\n")
+
+    private fun builderRequiredMethodFlagsDefinitions() = configMethods.filter { it.isRequired }.map {
+        writable { rust("pub(crate) ${it.requiredBuilderFlagName()}: bool,") }
+    }.join("\n")
+
+    private fun builderRequiredMethodFlagsInit() = configMethods.filter { it.isRequired }.map {
+        writable { rust("${it.requiredBuilderFlagName()}: false,") }
+    }.join("\n")
+
+    private fun builderRequiredMethodFlagsMove() = configMethods.filter { it.isRequired }.map {
+        writable { rust("${it.requiredBuilderFlagName()}: self.${it.requiredBuilderFlagName()},") }
+    }.join("\n")
+
+    private fun builderRequiredMethodError() = writable {
+        val variants = configMethods.filter { it.isRequired }.map {
+            writable {
+                rust(
+                    """
+                    ##[error("service is not fully configured; invoke `${it.requiredBuilderFlagName()}` on the config builder")]
```

Comment:
Wouldn't `${it.requiredBuilderFlagName()}` output the name of the flag itself, rather than the method name that the user is expected to call on `Config::builder`?

Reply:
Good catch!